src/scripts/gap_detector.py

```diff
@@ -0,0 +1,147 @@
+
+    def __init__(self, dune: DuneClient, database: MultiInstanceDBFetcher):
+        self.dune = dune
+        self.database = database
+
+    def dune_df(self, query_id: int, start: int, end: int) -> DataFrame:
+        """Executes and fetches dataframe from Dune `query_id`"""
+        return pd.read_csv(
+            self.dune.refresh_csv(
+                Query(
+                    query_id=query_id,
+                    params=[
+                        QueryParameter.number_type("start", start),
+                        QueryParameter.number_type("end", end),
+                    ],
+                )
+            ).data
+        )
+
+    def db_df(self, query_str: str, start: int, end: int) -> DataFrame:
+        """Executes and fetches dataframe for `query_str` from Database"""
+        return self.database.exec_query(
+            query_str.replace("{{start}}", str(start)).replace("{{end}}", str(end)),
+            self.database.connections[0],
+        )
+
+    def tx_list_diff(
+        self,
+        start: int,
+        end: int,
+    ) -> set[str]:
+        """Gets the list of differing transaction hashes between Dune and Database"""
+        dune_hashes = set(self.dune_df(DUNE_HASH_QUERY_ID, start, end)["tx_hash"])
+        db_hashes = set(self.db_df(DB_HASH_QUERY, start, end)["tx_hash"])
+        # TODO - usually its the database missing something (i.e. we assume Dune contains all).
+        #  However, we could also return an object that keeps track of what is missing where.
+        return dune_hashes.symmetric_difference(db_hashes)
+
+    def count_diff(
+        self,
+        start: int,
+        end: int,
+    ) -> int:
+        """
+        This function acts as a quick comparison between data sources for a block range.
+        The assumption is that if the number of batches between a certain block range agrees,
+        between the two sources then they are correct. If a difference is detected, one could
+        drill deeper and compare the set of transaction hashes. The transaction hash comparison
+        would also reveal any other types of disagreement that the count does not catch.
+        """
+        dune_count = self.dune_df(DUNE_COUNT_QUERY_ID, start, end)
+        db_count = self.db_df(DB_COUNT_QUERY, start, end)
+
+        return int(dune_count["batches"][0]) - int(db_count["batches"][0])
+
+    def find_missing(
+        self,
+        start: int,
+        end: int,
+    ) -> set[str]:
+        """
+        A recursive binary search, returning any disagreement
+        in transaction hash sets (between dune and database) for a given block range.
+        """
+        print("Inspecting Block Range...", start, end)
+        count_diff = self.count_diff(start, end)
+        if count_diff == 0:
+            # Nothing missing here!
+            print("counts agree on", start, end)
+            return set()
+
+        if count_diff < 200:  # TODO - make this configurable.
```

Comment:
Nit: Same as above. Generate an issue and reference in todo.

Reply:
Actually I am considering changing this to be `block_range_diff`, but I will create an issue for sure